shared/health/src/main/java/io/pravega/shared/health/impl/CompositeHealthContributor.java

```diff
@@ -0,0 +1,121 @@
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Objects;
+
+/**
+ * The {@link CompositeHealthContributor} class defines the base logic required to build {@link HealthContributor} objects
+ * that depend on (or are composed of) other {@link HealthContributor} objects. To do this we need to due two things:
+ * reducing/reconciling the (potentially) many different {@link Status} states and a way to persist these relationships.
+ *
+ * {@link CompositeHealthContributor} objects are used *strictly* for grouping, meaning that they do not define their own
+ * health checking logic and therefore does not export any details ({@link Health#getDetails}).
+ */
+@Slf4j
+@ThreadSafe
+public abstract class CompositeHealthContributor implements HealthContributor {
+
+    /**
+     * The {@link StatusAggregator} used to perform the aggregation of all the {@link HealthContributor} dependencies.
+     */
+    @Getter
+    private final StatusAggregator aggregator;
+
+    private final ContributorRegistry registry;
+
+    private final Collection<HealthContributor> contributors = ImmutableSet.of();
+
+    CompositeHealthContributor(StatusAggregator aggregator, ContributorRegistry registry) {
+        this.aggregator = aggregator;
+        this.registry = registry;
+    }
+
+    /**
+     * Requests the {@link Health} without fetching any details from it's dependencies.
+     * @return The {@link Health} result.
+     */
+    public Health getHealthSnapshot() {
+        return getHealthSnapshot(false);
+    }
+
+    /**
+     * Performs a {link Health} check on *all* descendant {@link HealthContributor}. The 'includeDetails' flag
+     * has different semantics depending on the type of sub-class.
+     *
+     * {@link io.pravega.shared.health.HealthIndicator} object will be requested to fetch their {@link Health#getDetails()}
+     * while {@link HealthComponent} will provide the {@link Health} results for all it's descendants.
+     *
+     * @param includeDetails Whether or not to request details from each of the {@link HealthContributor}.
+     * @return The {@link Health} result of the {@link CompositeHealthContributor}.
+     */
+    public Health getHealthSnapshot(boolean includeDetails) {
+        Health.HealthBuilder builder = Health.builder().name(getName());
+        Collection<Status> statuses = new ArrayList<>();
+        // Fetch the Health Status of all dependencies.
+        val children =  contributors().stream()
+                .filter(Objects::nonNull)
+                .map(contributor -> {
+                    Health health = contributor.getHealthSnapshot(includeDetails);
+                    statuses.add(health.getStatus());
+                    if (health.getStatus() == Status.UNKNOWN) {
+                        log.warn("{} has a Status of 'UNKNOWN'. This indicates `doHealthCheck` does not set a status" +
```

Comment:
Would this be logged every time we call `getHealthSnapshot`? I really do not like dumping log warnings every time we invoke a method like this. Is there another way to handle this?

Reply:
That's correct. If you think this is too verbose I think we can rely on this effect being noticed anyways simply through how this status may affect any parent statuses (e.g. it is apart of an aggregate).